### DISCUSSION TITLE
Throwing SyntaxError instead of Error

### DISCUSSION THREAD

**Discussion by idan-at** [swc-project/swc#2650] [General]
Hey,

when using `@swc/core` and one of the `transform*` methods, In case of a syntax error, it throws a very detailed error, but of type Error with a `GenericFailure` code.

I think it is more appropriate to throw the standard `SyntaxError` (which babel also throws on these cases).

WDYT? Is there a particular reason swc does not use `SyntaxError` ?

Thanks!

**Comment by kdy1** [comment]
Seems like a good change to me.
There's no special reason to use `Error`.

**Reply by idan-at** [reply]
can you refer me? I'd be happy to contribute.

**Reply by kdy1** [reply]
https://github.com/swc-project/swc/blob/a6836368a2dd466e20e41705811760bdfc259e4c/node/binding/src/util.rs#L35-L38

Check for if the error contains a cause with message `Syntax Error` will work.

**Reply by idan-at** [reply]
@kdy1 thanks for your previous response.

The relevant functions didn't exist on N-API. I've added those to node and also to `napi-rs`, [under](https://github.com/napi-rs/napi-rs/pull/897) the `experimental` flag.

I see `swc` still uses v1. Are you planning to upgrade soon?
Also, from my understanding, `napi-rs` needs to expose an `experimental` feature flag [here](https://github.com/napi-rs/napi-rs/blob/main/crates/napi/Cargo.toml), so `swc` can declare using it?

**Reply by idan-at** [reply]
@kdy1 I've [contributed to napi-rs](https://github.com/napi-rs/napi-rs/pull/897) and I see swc was recently updated to use napi-rs v2.

Can you please LMK if it's fine to add the `experimental` flag and use the new APIs?

Thanks.

**Reply by kdy1** [reply]
I think a flag is not required.
All `SyntaxErorr` is `Error` and the only way a user can match it is looking at the error message.

**Reply by idan-at** [reply]
I mean adding the feature flag [here](https://github.com/swc-project/swc/blob/206da128a1f1ecd43c7580a7ad5f58ebc363c812/crates/node/Cargo.toml#L25)

Also, I see that the APIs were only added in node 17.2.0. Can we somehow add a flag to check which version swc is running with and perform this change only if the functions exists?

Sorry for the noob questions, I'm new to the Rust environment.

**Reply by kdy1** [reply]
Oh.
> Also, I see that the APIs were only added in node 17.2.0. Can we somehow add a flag to check which version swc is running with and perform this change only if the functions exists?

This seems like a serious issue.
I'm not sure if `napi-rs` supports such behavior
